snapcraft/models/project.py

```diff
@@ -436,13 +437,36 @@ def _validate_autostart_name(cls, name):
 
         return name
 
-    @pydantic.field_validator("bus_name")
+    @pydantic.field_validator(
+        "command", "stop_command", "post_stop_command", "reload_command", "bus_name"
+    )
     @classmethod
-    def _validate_bus_name(cls, name):
-        if not re.match(r"^[A-Za-z0-9/. _#:$-]*$", name):
-            raise ValueError(f"{name!r} is not a valid bus name")
+    def _validate_apps_section_content(cls, command: str, info: pydantic.ValidationInfo) -> str:
+        # Find any invalid characters in the command.
+        # The regex below is derived from snapd's validator code, modified to be the inverse (^).
+        # https://github.com/canonical/snapd/blob/0706e2d0b20ae2bf030863f142b8491b66e80bcb/snap/validate.go#L756
+        bad_chars: set[str] = set()
+        for bad_char in re.finditer(r"[^A-Za-z0-9/. _#:$-]", command):
```

Comment:
Why not just set the attributes' class to be annotated strings with the regex match?

Reply:
I like this semantically, but I don't see a way to control the error message that gets emitted upon validation failure. Thus, I don't think we should do this.

```py
class Foo(pydantic.BaseModel):
    s: typing.Annotated[str, pydantic.Field(pattern=r"a")]

b = Foo(s="b")
```

```
Traceback (most recent call last):
  File "/home/imani/canonical/snapcraft/test.py", line 8, in <module>
    b = Foo(s="b")
  File "/home/imani/canonical/snapcraft/.venv/lib/python3.10/site-packages/pydantic/main.py", line 212, in __init__
    validated_self = self.__pydantic_validator__.validate_python(data, self_instance=self)
pydantic_core._pydantic_core.ValidationError: 1 validation error for Foo
s
  String should match pattern 'a' [type=string_pattern_mismatch, input_value='b', input_type=str]
    For further information visit https://errors.pydantic.dev/2.9/v/string_pattern_mismatch
```